ts/ast/ast.go

```diff
@@ -0,0 +1,272 @@
+package ast
+
+import (
+	"fmt"
+	"strings"
+)
+
+const Indent = "  "
+
+type File struct {
+	Nodes []Node
+}
+
+func (f File) String() string {
+	var b strings.Builder
+
+	for i, n := range f.Nodes {
+		b.WriteString(n.String())
+
+		if i+1 < len(f.Nodes) {
+			b.WriteString("\n\n")
+		}
+	}
+	b.WriteString("\n")
+
+	return b.String()
+}
+
+type Node interface {
+	fmt.Stringer
+}
+
+type Expr interface {
+	Node
+	expr()
+}
+
+type Decl interface {
+	Node
+	decl()
+}
+
+var (
+	_ Expr = SelectorExpr{}
+	_ Expr = IndexExpr{}
+	_ Expr = Num{}
+)
+
+type Raw struct {
+	Data string
+}
+
+func (r Raw) decl() {}
+func (r Raw) expr() {}
+func (r Raw) String() string {
+	return r.Data
+}
+
+type Ident struct {
+	Name string
+}
+
+func (i Ident) expr() {}
+func (i Ident) String() string {
+	return i.Name
+}
+
+func None() Expr {
+	return Ident{}
+}
+
+type SelectorExpr struct {
+	Expr Expr
+	Sel  Ident
+}
+
+func (s SelectorExpr) expr() {}
+func (s SelectorExpr) String() string {
+	return fmt.Sprintf("%s.%s", s.Expr, s.Sel)
+}
+
+type IndexExpr struct {
+	Expr  Expr
+	Index Expr
+}
+
+func (i IndexExpr) expr() {}
+func (i IndexExpr) String() string {
+	return fmt.Sprintf("%s[%s]", i.Expr, i.Index)
+}
+
+type AssignExpr struct {
+	Name  Ident
+	Value Expr
+}
+
+func (a AssignExpr) expr() {}
+func (a AssignExpr) String() string {
+	return fmt.Sprintf("%s = %s", a.Name, a.Value)
+}
+
+type KeyValueExpr struct {
+	Key   Expr
+	Value Expr
+}
+
+func (k KeyValueExpr) expr() {}
+func (k KeyValueExpr) String() string {
+	return fmt.Sprintf("%s: %s", k.Key, k.Value)
+}
+
+type UnaryExpr struct {
+	Op   string // operator
+	Expr Expr   // operand
+}
+
+func (u UnaryExpr) expr() {}
+func (u UnaryExpr) String() string {
+	return u.Op + u.Expr.String()
+}
+
+type BinaryExpr struct {
+	Op   string
+	X, Y Expr
+}
+
+func (b BinaryExpr) expr() {}
+func (b BinaryExpr) String() string {
+	return fmt.Sprintf("%s %s %s", b.X, b.Op, b.Y)
+}
+
+type Num struct {
+	N   interface{}
+	Fmt string
+}
+
+func (n Num) expr() {}
+func (n Num) String() string {
+	if n.Fmt == "" {
+		return fmt.Sprintf("%v", n.N)
+	}
+	return fmt.Sprintf(n.Fmt, n.N)
+}
+
+type Str struct {
+	Value string
+}
+
+func (s Str) expr() {}
+func (s Str) String() string {
+	return fmt.Sprintf(`'%s'`, s.Value)
+}
+
+type ObjectLit struct {
+	Elems []KeyValueExpr
+}
+
+func (o ObjectLit) expr() {}
+func (o ObjectLit) String() string {
+	var b strings.Builder
+	b.WriteString("{\n")
+	for _, e := range o.Elems {
+		b.WriteString(Indent)
+		b.WriteString(e.String())
+		b.WriteString(",\n")
+	}
+	b.WriteString("}")
+	return b.String()
+}
+
+type VarDecl struct {
+	Tok string
+
+	Name  Ident
+	Type  Ident
+	Value Expr
+}
+
+func (v VarDecl) decl() {}
+func (v VarDecl) String() string {
+	tok := v.Tok
+	if tok == "" {
+		tok = "const"
+	}
+	return fmt.Sprintf("%s %s: %s = %s;", tok, v.Name, v.Type, v.Value)
+}
+
+type Type interface {
```

Comment:
My brain is stuttering over all the use of `Type`. Can we use `Kind` instead, particularly for consistency with the attribute names?